core/src/main/java/com/netflix/msl/msg/MessageBuilder.java

```diff
@@ -604,45 +622,122 @@ public MessageBuilder addServiceToken(final ServiceToken serviceToken) throws Ms
      *         built.
      */
     public MessageBuilder addServiceTokenIfAbsent(final ServiceToken serviceToken) throws MslMessageException {
-        if (!serviceTokens.containsKey(serviceToken.getName()))
-            addServiceToken(serviceToken);
+        final Iterator<ServiceToken> tokens = serviceTokens.iterator();
+        while (tokens.hasNext()) {
+            final ServiceToken token = tokens.next();
+            if (token.getName().equals(serviceToken.getName()) &&
+                token.isMasterTokenBound() == serviceToken.isMasterTokenBound() &&
+                token.isUserIdTokenBound() == serviceToken.isUserIdTokenBound())
+            {
+                return this;
+            }
+        }
+        addServiceToken(serviceToken);
         return this;
     }
+    
+    /**
+     * <p>Exclude a service token from the message. This matches the token name
+     * and whether or not it is bound to the master token or to a user ID
+     * token. It does not require the token to be bound to the exact same
+     * master token or user ID token that will be used in the message.</p>
+     * 
+     * <p>The service token will not be sent in the built message. This is not
+     * the same as requesting the remote entity delete a service token.</p>
+     * 
+     * <p>This function is equivalent to calling
+     * {@link #excludeServiceToken(String, boolean, boolean)}.</p>
+     * 
+     * @param serviceToken the service token.
+     * @return this.
+     * @see #excludeServiceToken(String, boolean, boolean)
+     */
+    public MessageBuilder excludeServiceToken(final ServiceToken serviceToken) {
+        return excludeServiceToken(serviceToken.getName(), serviceToken.isMasterTokenBound(), serviceToken.isUserIdTokenBound());
+    }
 
     /**
-     * <p>Exclude a service token from the message.</p>
+     * <p>Exclude a service token from the message matching all the specified
+     * parameters. A false value for the master token bound or user ID token
+     * bound parameters restricts exclusion to tokens that are not bound to a
+     * master token or not bound to a user ID token respectively.</p>
+     * 
+     * <p>For example, if a name is provided and the master token bound
```

Comment:
What happens if you set masterTokenBound=false and userIdTokenBound=true?

Reply:
Nothing, from looking at the code. But it bothers me that you can call this way since that binding is not allowed. Might be worth a comment.

Reply:
Clarified function documentation.